File: src/simulation/gunnery.ts
import { surfaceGunAllowed } from '../ships/armament';
import { selectedWeapon, weaponGroupId } from '../ships/weaponGroups';
import type { Ammunition, Battery, Vec3 } from '../ships/blueprint';
import type { AirContext, Aircraft } from './aircraft';
import { antiAircraftCandidates, updateAntiAircraft } from './antiAircraft';
import type { FleetActor } from './battle';
import { updateMountCarrier } from './mountFrames';
import { botAim, botAmmunition, botDidFire, botGunRange, botReadyToFire, clearFiringLane, shipVelocity } from './bots';
import { dispersedDirection, dispersedSpeed, velocityPenetration } from './ballistics';
import { add, length, scale } from './geometry';
import { directorDispersion, equipmentCondition, mountSupport, supportPerformance } from './machinery';
import { availableAmmunition, expendSalvo, muzzleWorld, queueAmmunition, selectAmmunition, shotDirection, updateMount, type MountDefinition, type MountState } from './weapons';

/** The player's intent for one tick. Only the selected group follows the sight;
 * every other surface gun stays idle and every registered AA gun stays automatic. */
export interface PlayerGunOrders {
  battery: Battery;
  weaponGroupId?: string;
  /** Undefined when the host supplied a malformed sight; selected guns then report out-of-arc. */
  aim?: Vec3;
  fire: boolean;
  /** Ordered shell type per weapon group id, with the battery as fallback. */
  ammunition: Readonly<Record<string, Ammunition>>;
}

export interface GunneryContext extends AirContext {
  seed: number;
  dt: number;
  /** Seeded surface dispersion advances per shell, independently of projectile ids. */
  nextDispersion(): number;
  /** This tick's airborne, off-deck aircraft in stable order. Undefined outside an
   * active battle, where no gun engages aircraft. */
  airborne?: readonly Aircraft[];
}

/** Operate every gun mount of one ship for a fixed tick. Each mount advances its
 * reload, recoil and training exactly once. Automatic air defense claims a mount
 * first unless the player has selected its group directly; otherwise the mount
 * follows the sight, its bot's track, or simply keeps reloading. A salvo spends
 * one complete load from the stock shared by both firing paths. */
export function operateGuns(actor: FleetActor, ctx: GunneryContext, target?: FleetActor, player?: PlayerGunOrders): void {
  const def = actor.definition;
  const aircraft = ctx.airborne && antiAircraftCandidates(actor, ctx.airborne);
  const support = supportPerformance(actor, def);
  const laneClear = !!target && clearFiringLane(actor, target, ctx.actors);
  const velocity = shipVelocity(actor);
  def.mounts.forEach((m, i) => {
    updateMountCarrier(def, i, actor.mounts);
    const state = actor.mounts[i];
    if (actor.damage.stability.combatLost) { state.status = 'disabled'; return; }
    if (m.magazineId && equipmentCondition(actor, def, m.magazineId).availability === 0) { state.status = 'disabled'; return; }
    const surfaceAllowed = surfaceGunAllowed(def, m.weapon);
    const selected = surfaceAllowed && !!player && selectedWeapon(m.battery, m.weapon, player.battery, player.weaponGroupId);
    // A directly selected group belongs to the sight; aggregate battery intent keeps AA automatic.
    const manual = selected && player!.weaponGroupId !== undefined;
    if (aircraft && !manual && updateAntiAircraft(actor, m, state, ctx, ctx.dt, aircraft, support.power)) return;
    if (!surfaceAllowed) { updateMount(m, state, def, actor.motion, undefined, ctx.dt, velocity, support.power, actor.mounts); return; }
    let aim: Vec3 | undefined, fire = false, inRange = false;
    if (player) {
      queueAmmunition(m, state, player.ammunition[weaponGroupId(m.battery, m.weapon)] ?? player.ammunition[m.battery]);
      if (!player.aim) { state.status = 'out-of-arc'; return; }
      aim = player.aim; fire = player.fire && selected;
    } else if (actor.controller === 'bot' && target) {
      selectAmmunition(m, state, botAmmunition(target, m, state));
      inRange = Math.hypot(target.motion.x - actor.motion.x, target.motion.z - actor.motion.z) <= botGunRange(m);
      if (inRange && state.hp > 0 && availableAmmunition(state) >= (m.weapon.barrelCount ?? 2)) aim = botAim(actor, target, m, state);
      fire = inRange && laneClear && botReadyToFire(actor, m);
    }
    const aligned = updateMount(m, state, def, actor.motion, aim, ctx.dt, velocity, support.power, actor.mounts);
    if (actor.damage.sunk || !fire || !aligned || state.status !== 'ready') return;
    if (actor.controller === 'bot') botDidFire(actor, m);
    fireSurfaceSalvo(actor, m, state, support.power, velocity, ctx);
  });
}

/** Spend the loaded salvo and launch one dispersed shell per barrel from its own muzzle. */
function fireSurfaceSalvo(actor: FleetActor, m: MountDefinition, state: MountState, power: number, velocity: Vec3, ctx: GunneryContext): void {
  const def = actor.definition, w = m.weapon;
  const barrels = expendSalvo(m, state);
  const spread = (w.ballistics?.dispersionRad ?? 0) + directorDispersion(mountSupport(actor, def, m.id, power).fireControl);
  const label = `${Math.round(w.caliberM * 1000)} mm ${state.loaded.toUpperCase()} · ${m.battery === 'main' ? 'Main' : 'Secondary'}`;
  for (let barrel = 0; barrel < barrels; barrel++) {
    const position = muzzleWorld(m, state, barrel, actor.motion);
    const shot = ctx.nextDispersion();
    const direction = dispersedDirection(shotDirection(m, state, actor.motion), spread, ctx.seed, shot);
    const speed = dispersedSpeed(w.muzzleSpeed, w.ballistics?.muzzleSpeedSigmaFraction ?? 0, ctx.seed, shot);
    const shellVelocity = add(scale(direction, speed), velocity);
    const id = ctx.nextId();
    ctx.shells.push({ id, ownerId: actor.motion.id, weaponLabel: label, position, velocity: shellVelocity, age: 0,
      penetrationMm: state.loaded === 'he' ? 0 : velocityPenetration(w.penetrationMm, w.ballistics?.penetrationReferenceSpeedMps ?? w.muzzleSpeed, length(shellVelocity)),
      damage: w.damage, caliberM: w.caliberM, visited: [], ammunition: state.loaded,
      ap: state.loaded === 'ap' ? w.ap : undefined, he: state.loaded === 'he' ? w.he : undefined, dragPerSecond: w.ballistics?.dragPerSecond ?? 0 });
    ctx.emit({ kind: 'shot', position: [...position], shipId: actor.motion.id, message: `${m.name} fired`,
      shell: { id, caliberM: w.caliberM, velocity: [...shellVelocity], ammunition: state.loaded, type: state.loaded === 'he' ? 'HE' : 'AP' } });
  }
}

File: crates/naval-sim/src/gunnery.rs
use crate::{
    anti_aircraft,
    aviation::Aviation,
    ballistics::*,
    bots,
    definition::{MountDefinition, Vec3},
    geometry::*,
    impact::{DamageEvent, ShellEffect},
    machinery::{electrical_power, equipment_condition, mount_support},
    shell::Shell,
    vessel::{Controller, Vessel},
    weapons::*,
};
use std::collections::BTreeMap;
#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerGunOrders {
    pub battery: String,
    pub weapon_group_id: Option<String>,
    pub aim: Option<Vec3>,
    pub fire: bool,
    pub ammunition: BTreeMap<String, Ammunition>,
}
/// Group identity derives from fitted weapon characteristics, like the authoring
/// adapter. Normalize integral JSON numbers to JavaScript's stable group keys.
pub fn group_id(m: &MountDefinition) -> String {
    let w = &m.weapon;
    let b = w.ballistics.as_ref();
    let mut fields = serde_json::json!([
        w.caliber_m,
        w.reload_seconds,
        w.muzzle_speed,
        w.projectile_mass_kg,
        w.penetration_mm,
        w.damage,
        w.traverse_rate_deg,
        w.elevation_min_deg,
        w.elevation_max_deg,
        w.elevation_rate_deg,
        b.map_or(0.0, |b| b.drag_per_second),
        b.map_or(0.0, |b| b.dispersion_rad),
        b.map_or(0.0, |b| b.muzzle_speed_sigma_fraction.unwrap_or(0.0)),
        b.map_or(w.muzzle_speed, |b| b
            .penetration_reference_speed_mps
            .unwrap_or(w.muzzle_speed)),
        w.ap.as_ref().map(|a| [
            a.arming_resistance_mm,
            a.fuze_delay_seconds,
            a.explosive_kg,
            a.fragment_penetration_mm
        ]),
        w.he.as_ref()
            .map(|h| [h.explosive_kg, h.fragment_penetration_mm, h.damage])
    ]);
    fn normalize(v: &mut serde_json::Value) {
        match v {
            serde_json::Value::Array(a) => a.iter_mut().for_each(normalize),
            serde_json::Value::Number(n) => {
                let f = n.as_f64().unwrap();
                if f.fract() == 0.0 && f.abs() < 9e15 {
                    *v = serde_json::json!(f as i64)
                }
            }
            _ => (),
        }
    }
    normalize(&mut fields);
    format!("{}:gun:{}", m.battery, fields)
}
pub struct GunneryContext<'a> {
    pub actors: &'a [Vessel],
    pub aviation: &'a mut Aviation,
    pub shells: &'a mut Vec<Shell>,
    pub sequence: &'a mut i64,
    pub dispersion: &'a mut u32,
    pub events: &'a mut Vec<DamageEvent>,
    pub seed: u32,
    pub dt: f64,
}
/// Each mount is temporarily detached, avoiding fleet clones and mutable aliases
/// while both automatic and manual paths inspect the owning vessel.
pub fn operate(
    actor: &mut Vessel,
    ctx: &mut GunneryContext<'_>,
    target: Option<&Vessel>,
    player: Option<&PlayerGunOrders>,
) {
    let compiled = actor.compiled.clone();
    let def = &compiled.definition;
    let power = electrical_power(actor, def, None);
    let lane = target.is_some_and(|t| bots::clear_firing_lane(actor, t, ctx.actors));
    let velocity = actor.motion.velocity();
    for (i, m) in def.mounts.iter().enumerate() {
        if actor.damage.stability.combat_lost
            || m.magazine_id.as_ref().is_some_and(|id| {
                def.modules
                    .iter()
                    .find(|m| &m.id == id)
                    .is_some_and(|module| {
                        equipment_condition(actor, def, module, None).availability == 0.0
                    })
            })
        {
            actor.mounts[i].status = "disabled".into();
            continue;
        }
        let allowed = anti_aircraft::surface_allowed(def, m);
        let group = &compiled.weapon_group_ids[i];
        let selected = allowed
            && player.is_some_and(|p| {
                p.battery == m.battery && p.weapon_group_id.as_ref().is_none_or(|id| id == group)
            });
        let manual = selected && player.unwrap().weapon_group_id.is_some();
        let mut state = actor.mounts[i].clone();
        if !manual
            && anti_aircraft::update(
                actor,
                m,
                &mut state,
                ctx.actors,
                ctx.aviation,
                ctx.dt,
                ctx.seed,
                ctx.sequence,
                ctx.events,
            )
        {
            actor.mounts[i] = state;
            continue;
        }
        if !allowed {
            update_mount(
                m,
                &mut state,
                def,
                &actor.motion,
                None,
                ctx.dt,
                velocity,
                power,
                &compiled.obstructions,
            );
            actor.mounts[i] = state;
            continue;
        }
        let mut aim = None;
        let mut fire = false;
        if let Some(p) = player {
            if let Some(kind) = p
                .ammunition
                .get(group)
                .or_else(|| p.ammunition.get(&m.battery))
            {
                state.queue_ammunition(m, *kind)
            }
            let Some(point) = p.aim else {
                state.status = "out-of-arc".into();
                actor.mounts[i] = state;
                continue;
            };
            aim = Some(point);
            fire = p.fire && selected;
        } else if actor.controller == Controller::Bot
            && let Some(t) = target
        {
            state.select_ammunition(m, bots::ammunition(t.definition(), m, &state));
            let in_range = (t.motion.x - actor.motion.x).hypot(t.motion.z - actor.motion.z)
                <= bots::gun_range(m);
            if in_range
                && state.hp > 0.0
                && state.available(state.loaded) >= m.weapon.barrel_count.unwrap_or(2.0)
            {
                aim = Some(bots::aim(
                    actor.bot.as_ref(),
                    &actor.motion,
                    t,
                    t.definition(),
                    m,
                    &mut state,
                ))
            }
            fire = in_range && lane && actor.bot.as_ref().is_some_and(|b| b.ready(Some(m)));
        }
        let aligned = update_mount(
            m,
            &mut state,
            def,
            &actor.motion,
            aim,
            ctx.dt,
            velocity,
            power,
            &compiled.obstructions,
        );
        if !actor.damage.sunk && fire && aligned && state.status == "ready" {
            if actor.controller == Controller::Bot
                && let Some(bot) = actor.bot.as_mut()
            {
                bot.did_fire(m)
            }
            let barrels = state.expend_salvo(m, m.weapon.reload_seconds);
            let w = &m.weapon;
            let spread = w.ballistics.as_ref().map_or(0.0, |b| b.dispersion_rad)
                + (1.0 - mount_support(actor, def, Some(&m.id), None).1) * 0.0015;
            let ammo = if state.loaded == Ammunition::He {
                "HE"
            } else {
                "AP"
            };
            let label = format!(
                "{} mm {ammo} · {}",
                (w.caliber_m * 1000.0).round() as i64,
                if m.battery == "main" {
                    "Main"
                } else {
                    "Secondary"
                }
            );
            for barrel in 0..barrels {
                let position = local_to_world(
                    muzzle_local(m, state.train, state.elevation, barrel),
                    actor.motion.pose(),
                );
                let shot = *ctx.dispersion;
                *ctx.dispersion = ctx.dispersion.wrapping_add(1);
                let direction = dispersed_direction(
                    shot_direction(m, &state, actor.motion.pose()),
                    spread,
                    ctx.seed,
                    shot,
                );
                let speed = dispersed_speed(
                    w.muzzle_speed,
                    w.ballistics
                        .as_ref()
                        .map_or(0.0, |b| b.muzzle_speed_sigma_fraction.unwrap_or(0.0)),
                    ctx.seed,
                    shot,
                );
                let shell_velocity = add(scale(direction, speed), velocity);
                *ctx.sequence += 1;
                let shell = Shell {
                    id: *ctx.sequence,
                    owner_id: actor.motion.id.clone(),
                    weapon_label: Some(label.clone()),
                    position,
                    velocity: shell_velocity,
                    penetration_mm: if state.loaded == Ammunition::He {
                        0.0
                    } else {
                        velocity_penetration(
                            w.penetration_mm,
                            w.ballistics.as_ref().map_or(w.muzzle_speed, |b| {
                                b.penetration_reference_speed_mps.unwrap_or(w.muzzle_speed)
                            }),
                            length(shell_velocity),
                        )
                    },
                    damage: w.damage,
                    caliber_m: w.caliber_m,
                    ammunition: Some(state.loaded),
                    ap: if state.loaded == Ammunition::Ap {
                        w.ap.clone()
                    } else {
                        None
                    },
                    he: if state.loaded == Ammunition::He {
                        w.he.clone()
                    } else {
                        None
                    },
                    drag_per_second: Some(w.ballistics.as_ref().map_or(0.0, |b| b.drag_per_second)),
                    ..Default::default()
                };
                ctx.events.push(DamageEvent {
                    kind: "shot".into(),
                    position,
                    ship_id: actor.motion.id.clone(),
                    message: format!("{} fired", m.name),
                    shell: Some(ShellEffect::from_shell(&shell)),
                    ..Default::default()
                });
                ctx.shells.push(shell);
            }
        }
        actor.mounts[i] = state;
    }
}
